simulatingrisk/hawkdovemulti/batch_run.py

```diff
@@ -0,0 +1,266 @@
+    datestr = datetime.today().isoformat().replace(".", "_").replace(":", "")
+    model_output_filename = os.path.join(data_dir, f"{file_prefix}{datestr}_model.csv")
+    if collect_agent_data:
+        agent_output_filename = os.path.join(
+            data_dir, f"{file_prefix}{datestr}_agent.csv"
+        )
+
+    message = f"Saving data collection results to:\n  {model_output_filename}"
+    if collect_agent_data:
+        message += f"\n  {agent_output_filename}"
+    print(message)
+
+    # open output files so data can be written as it is generated
+    with open(model_output_filename, "w", newline="") as model_output_file:
+        if collect_agent_data:
+            agent_output_file = open(agent_output_filename, "w", newline="")
+
+        model_dict_writer = None
+        agent_dict_writer = None
+
+        # adapted from mesa batch run code
+        with tqdm(total=total_runs, disable=not progressbar) as pbar:
+            with multiprocessing.Pool(number_processes) as pool:
+                for model_data, agent_data in pool.imap_unordered(
+                    run_hawkdovemulti_model, runs_list
+                ):
+                    # initialize dictwriter and start csv after the first batch
+                    if model_dict_writer is None:
+                        # get field names from first entry
+                        model_dict_writer = csv.DictWriter(
+                            model_output_file, model_data.keys()
+                        )
+                        model_dict_writer.writeheader()
+
+                    model_dict_writer.writerow(model_data)
+
+                    if collect_agent_data:
+                        if agent_dict_writer is None:
+                            # get field names from first entry
+                            agent_dict_writer = csv.DictWriter(
+                                agent_output_file, agent_data[0].keys()
+                            )
+                            agent_dict_writer.writeheader()
+
+                        agent_dict_writer.writerows(agent_data)
+
+                    pbar.update()
+
+        if collect_agent_data:
+            agent_output_file.close()
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        prog="hawk/dove batch_run",
+        description="Batch run for hawk/dove multiple risk attitude simulation.",
+        epilog="""Data files will be created in data/hawkdovemulti/
+        relative to current path.""",
+    )
+    parser.add_argument(
+        "-i",
+        "--iterations",
+        type=int,
+        help="Number of iterations to run for each set of parameters "
+        + "(default: %(default)s)",
+        default=100,
+    )
+    parser.add_argument(
+        "-m",
+        "--max-steps",
+        help="Maximum steps to run simulations if they have not already "
+        + "converged (default: %(default)s)",
+        default=1000,  # new convergence logic seems to converge around 400
```

Comment:
would there be any benefit to tunable convergence conditions via model parameter(s)? Might be more relevant for the single/no risk adjustment convergence logic which may be of less importance..

I'm not sure if there is anything further that can be done with the `HawkDoveMulti` convergence condition. Is it possible for there to be 'temporary stability' where `num_agents_risk_changed == 0` for one risk adjustment and then some agents manage to adjust the next?